Ensure source is set on GameplayClockContainer.Reset()

smoogipoo opened the pull request:
As pointed out by @bdach in https://github.com/ppy/osu/issues/12541, makes sense to me.

smoogipoo committed:
Ensure source is set on reset

frenzibyte commented on osu.Game/Screens/Play/GameplayClockContainer.cs:
```diff
@@ -100,6 +100,7 @@ public virtual void Seek(double time)
         /// </summary>
         public virtual void Reset()
         {
+            ChangeSource(SourceClock);
```
better with the comment from the other usage:
```suggestion
            // Ensure that the source clock is set.
            ChangeSource(SourceClock);

```

bdach replied:
would rather factor out a one-liner method at that point if we're nitpicking to this degree.

peppy replied:
also make the comment mention why it possibly wouldn't be already set. looking at this PR in isolation makes me raise an eyebrow, so a comment may be good to have, explaining what this requirement comes from

bdach replied:
Have made a change that attempts to explain better. Please advise if it's any better.

bdach committed:
Attempt to explain source initialisation better

bdach commented:
While running tests locally I noticed that `TestSceneMasterGameplayClockContainer.TestElapseThenReset` was failing here, but apparently didn't fail CI, and doesn't fail on `master`. So I'm a bit suspicious of this diff now, even though I proposed this change myself. Needs more investigation, I think...

bdach committed:
Merge branch 'master' into gcc-change-source-on-reset

bdach committed:
Fix potential failure in `ensureSourceClockSet()`

`ensureSourceClockSet()` was intended to only run when the adjustable
source hasn't been set at all yet. As it turns out permitting it to run
unconditionally can break the state of the underlying interpolated
clock. This is caused by the following factors:

* While the decoupleable clock is running, its `CurrentTime` does not
  come from either the source clock, or the internal stopwatch; it is
  instead calculated using the base `InterpolatingFramedClock` logic.

* A source change of a decoupleable clock seeks the provided source
  clock to the decoupleable's current time.

* When an interpolating clock is seeked (decoupleable clock is also
  an interpolating one), its interpolation state
  (`{Last,Current}InterpolatedTime`) are reset to 0.

* If the interpolating clock determines that its current time is too
  far away from the source's time (which was set when the source is
  changed), it will ignore the source and instead continue to use
  its current time until the source clock has caught up.

Overall, the source change is not really necessary if a source is
already there. The only reason to ensure it was set was to make sure
the first seek of the gameplay clock wasn't performed in decoupled
mode. Therefore, add a guard to make sure the source is only set if
there isn't one already.

bdach commented:
I believe I have fixed the failures. The last commit message explains what was going on, but in short:

* while decoupleable clock is running, its current time doesn't come from either the source or the stopwatch - it comes from the base interpolating clock
* a source change of a decoupleable seeks the source clock also
* when an interpolating clock is seeked, its interpolation state (last/current interpolated time) are reset to 0
* that triggers [this branch](https://github.com/ppy/osu-framework/blob/master/osu.Framework/Timing/InterpolatingFramedClock.cs#L91-L95) in the interpolating clock, causing the current time of the interpolating clock to get stuck at the time at the point of the source change, until the framed source clock catches up.

I don't believe this would ever show up in a non-test usage (since in game `Reset()` is only ever used in the gameplay start process), but it *was* an actual issue if the reset was preceded by a `Start()` call. Not entirely sure why the failure didn't pop up on CI, but I've spent enough time on figuring this stuff out anyway for me to want to tap out for now.
